File: C++/ASP Advanced/Arrays/Kadane's Algorithm/Leetcode-53.cpp
#include <vector>
class Solution {
public:
    int maxSubArray(std::vector<int>& nums) {
        int totalSum = nums[0], sum = 0;

        for (int l = 0; l < nums.size(); l++){
            sum += nums[l];
            totalSum = std::max(totalSum, sum);
            if (sum < 0){
                sum = 0;
            } 
        }
        return totalSum;
    }
};
// The code defines a solution to the problem of finding the maximum sum of a contiguous subarray using Kadane's algorithm.
// It iterates through the array, maintaining a current sum and updating the maximum sum found so far.
// If the current sum becomes negative, it is reset to zero, as a negative sum would not contribute to a maximum subarray.

File: Python/ASP Advanced/Arrays/Kadane's Algorithm/Leetcode-53.py
# https://leetcode.com/problems/maximum-subarray/description/

class Solution:

    def maxSubArrayOptimal(self, nums: List[int]) -> int:
        maxSum, currSum = -1001, 0

        for num in nums:
            if currSum < 0:
                currSum = 0
            
            currSum += num
            maxSum = max(currSum, maxSum)

        return maxSum
    
    def maxSubArray(self, nums: List[int]) -> int:
        maxSum = nums[0]
        currSum = nums[0]
        for i in range(1, len(nums)):
            n = nums[i]

            if currSum + n < currSum and currSum + n < 0:
                currSum = n
            else:
                currSum = max(currSum + n, n)
                
            maxSum = max(currSum, maxSum)

        return maxSum